flow/raster_cache.cc

```diff
@@ -26,18 +26,32 @@ RasterCacheResult::RasterCacheResult(sk_sp<SkImage> image,
                                      const char* type)
     : image_(std::move(image)), logical_rect_(logical_rect), flow_(type) {}
 
-void RasterCacheResult::draw(SkCanvas& canvas, const SkPaint* paint) const {
+void RasterCacheResult::draw(SkCanvas& canvas,
+                             const SkPaint* paint,
+                             const SkMatrix* matrix,
+                             const SkSamplingOptions* sampling) const {
   TRACE_EVENT0("flutter", "RasterCacheResult::draw");
   SkAutoCanvasRestore auto_restore(&canvas, true);
-  SkIRect bounds =
-      RasterCache::GetDeviceBounds(logical_rect_, canvas.getTotalMatrix());
-  FML_DCHECK(
-      std::abs(bounds.size().width() - image_->dimensions().width()) <= 1 &&
-      std::abs(bounds.size().height() - image_->dimensions().height()) <= 1);
-  canvas.resetMatrix();
-  flow_.Step();
-  canvas.drawImage(image_, bounds.fLeft, bounds.fTop, SkSamplingOptions(),
-                   paint);
+  if (matrix == nullptr) {
+    SkIRect bounds =
+        RasterCache::GetDeviceBounds(logical_rect_, canvas.getTotalMatrix());
+    FML_DCHECK(
+        std::abs(bounds.size().width() - image_->dimensions().width()) <= 1 &&
+        std::abs(bounds.size().height() - image_->dimensions().height()) <= 1);
+    canvas.resetMatrix();
+    flow_.Step();
+    canvas.drawImage(image_, bounds.fLeft, bounds.fTop, SkSamplingOptions(),
+                     paint);
+  } else {
+    // Sampling options and matrix must be provided together.
+    FML_DCHECK(sampling != nullptr);
+    FML_DCHECK(matrix->isScaleTranslate());
+
+    SkRect dest_rect;
+    matrix->mapRect(&dest_rect, logical_rect_);
```

Comment:
mapRect sorts the points so if you have negative scales this will fail to produce the results you are looking for.

Reply:
Added a check for negative scales